src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp

```diff
@@ -0,0 +1,243 @@
+        std::swap(to_return, chunks);
+        return to_return;
+    }
+
+    /// Just read block is already enough.
+    if (isEnoughSize(chunk))
+    {
+        /// If no accumulated data, return just read block.
+        if (chunks.empty())
+        {
+            chunks.emplace_back(std::move(chunk));
+            return {};
+        }
+
+        /// Return accumulated data (maybe it has small size) and place new block to accumulated data.
+        std::vector<Chunk> to_return;
+        std::swap(to_return, chunks);
+        chunks.emplace_back(std::move(chunk));
+        return to_return;
+    }
+
+    /// Accumulated block is already enough.
+    if (isEnoughSize(chunks))
+    {
+        /// Return accumulated data and place new block to accumulated data.
+        std::vector<Chunk> to_return;
+        std::swap(to_return, chunks);
+        chunks.emplace_back(std::move(chunk));
+        return to_return;
+    }
+
+    chunks.emplace_back(std::move(chunk));
+    if (isEnoughSize(chunks))
+    {
+        std::vector<Chunk> to_return;
+        std::swap(to_return, chunks);
+        return to_return;
+    }
+
+    /// Squashed block is not ready.
+    return {};
+}
+
+std::pair<ColumnString::Ptr, ColumnString::Ptr> EmbeddedRocksDBBulkSink::serializeChunks(const std::vector<Chunk> & input_chunks) const
+{
+    auto serialized_key_column = ColumnString::create();
+    auto serialized_value_column = ColumnString::create();
+
+    {
+        auto & serialized_key_data = serialized_key_column->getChars();
+        auto & serialized_key_offsets = serialized_key_column->getOffsets();
+        auto & serialized_value_data = serialized_value_column->getChars();
+        auto & serialized_value_offsets = serialized_value_column->getOffsets();
+        WriteBufferFromVector<ColumnString::Chars> writer_key(serialized_key_data);
+        WriteBufferFromVector<ColumnString::Chars> writer_value(serialized_value_data);
+
+        for (const auto & chunk : input_chunks)
+        {
+            const auto & columns = chunk.getColumns();
+            auto rows = chunk.getNumRows();
+            for (size_t i = 0; i < rows; ++i)
+            {
+                for (size_t idx = 0; idx < columns.size(); ++idx)
+                    serializations[idx]->serializeBinary(*columns[idx], i, idx == primary_key_pos ? writer_key : writer_value, {});
+                writeChar('\0', writer_key);
+                writeChar('\0', writer_value);
+                serialized_key_offsets.emplace_back(writer_key.count());
+                serialized_value_offsets.emplace_back(writer_value.count());
+            }
+        }
+
+        writer_key.finalize();
+        writer_value.finalize();
+    }
+
+    return {std::move(serialized_key_column), std::move(serialized_value_column)};
+}
+
+void EmbeddedRocksDBBulkSink::consume(Chunk chunk_)
+{
+    std::vector<Chunk> to_written = squash(std::move(chunk_));
+
+    if (to_written.empty())
+        return;
```

Comment:
I don't think it can be called from multiple threads. If we have multiple threads for insert, it will create multiple `EmbeddedRocksDBBulkSink` objects. I will check more.

Reply:
Why `file_counter` is `std::atomic_size_t`, not `size_t` then?

Reply:
It should be a mistake, I will change to `size_t`.

Reply:
changed to `size_t`

Reply:
I recalled why it's `std::atomic_size` then. The inserting flow including two major steps: (1) create serialised key-value chunk and sort the chunk by key, and (2) create SST file and import it. (1) is compute-heavy, while (2) is I/O heavy, so I had an idea to overlap them. However, later I think that may be it's no different from having `max_insert_threads = 2`, so it's on hold now.